### ISSUE TITLE
Enforce-versions now disallows bugfix branches which branch from a hotfix.

### ISSUE THREAD

**Issue by jacobharris919** [egineering-llc/gitflow-helper-maven-plugin#87]
As of the newest plugin version (1.8.0), `enforce-versions` no longer allows this:

- Assume `master` is at version 1.0.0
- Branch `hotfix/1.0.1` from master  (POM version 1.0.1)
- Branch a correction from `hotfix/1.0.1` called `bugfix/good-bug-branch-description`  (POM version 1.0.1)

`enforce-versions` disallows the bugfix branch because it is now considered a "snapshot only" branch type.

I see that this change was intentional for some use cases, and it makes sense if the bugfix branch always comes off of `develop`, but I would like to see the new extra-restrictive behavior be an option you can turn on, with the default behavior allowing the same workflow that previous versions of the plugin allowed.  (i.e. `bugfix/` and `feature/` branches allowed to branch off of hotfix branches with non-snapshot versions)

The error looks like this:
```[ERROR] Failed to execute goal com.e-gineering:gitflow-helper-maven-plugin:1.8.0:enforce-versions (default) on project {project-name}: The current git branch: [origin/bugfix/good-bug-branch-description] is detected as a SNAPSHOT-type branch, and expects a maven project version ending with -SNAPSHOT. The maven project version found was: [1.0.1] -> [Help 1]```

**Comment by bvarner** [comment]
I think this originates from the changes in #83 -- which while valid in some cases certainly breaks the workflow of making bugfixes against a hotfix version.

I believe the intent of the enforcement here was to restrict overwriting of the hotfix artifact in the stage repo with a build from a branch yet approved for inclusion in the hotfix. @robth is that correct?

In thinking about this, I can see both sides of the argument, and this will affect features and bug-fixed being developed against a release branch as well.

### From the ivory-tower perspective that this enforcement was added from:
* The feature / bugfix branch should be a -SNAPSHOT version of the hotfix.

So in the example submitted with the issue report:
`bugfix/good-bug-branch-description` should have POM version 1.0.1-SNAPSHOT.

* You're working on something that shouldn't be deployed as the 1.0.1 artifact yet.
* Especially pertinent if you make a commit to this branch and trigger a deploy of 1.0.1 to the test repo right before someone actually merges the hotfix branch to master. You can sneak a change in.

### From the pragmatic "This will make pull requests hard" perspective:
* Foisting -SNAPSHOT on feature/bugfix branches is going to force some sort of pom.xml version mangling, or start failing CI builds before pull requests are completed. In order for the CI build to work the pom.xml will need to be -SNAPSHOT, but when you do the merge to the hotfix branch, you'll need to remove the -SNAPSHOT. 

That's a terrible workflow.

**Comment by bvarner** [comment]
Wouldn't it be nice if the plugin would generate the proper artifact version based off the one in the pom.xml and the current branch name and type?

**Comment by bvarner** [comment]
So with regards to this report, specifically: There's a clear regression in the function of the plugin as it was originally written, or there should be a switch to disable this enforcement on these branch types.

@jacobharris919 If you have some time, could you poke around and see if that would be a 'quick fix'?

Longer-term, I think there needs to be a re-visiting of the conversation around publishing of artifacts, and being able to identify the artifacts and where they were built from.

My leaning here is for non-release branches to append (part of) the source branch name to the classifier portion of the GAV for the project. This would resolve a few use-case issues that cropped up in #83 (which is why there's the branch-name-in the version enforcement), and would resolve issues with dynamic CD environments being able to spin up and identify the proper artifacts for the branch being deployed.

For example:
**master**
```
<groupId>foo</groupId>
<artifactId>bar</artifactId>
<version>1.0.0</version>
<packaging>jar</packaging>
```

There's an artifact in the release repo with `foo:bar:1.0.0:jar`.

**develop**
```
<version>1.1.0-SNAPSHOT</version>
```
Deploys artifacts to the snapshots repo with `foo:bar:1.1.0-SNAPSHOT:jar`.

**Hotfix (or release)**
```
<version>1.0.1</version>
```
Deploys artifacts to the test releases (stage) repo with `foo:bar:1.0.1:jar`.
Assume these are updated when a commit is pushed to /hotfix/1.0.1 and the build verifies.
The latest commit / tested code is in the repo as that artifact.

**Bugfix (or feature)** 
Should leave the version as-is from the branch it was created from.
Appends the branch name (or a portion of it) to a `<classifier/>` for the GAV.

* On branch `bugfix/barf` from `/hotfix/1.0.1`
    ```
    <version>1.0.1</version>
    ```
    Deploys artifacts to the test releases (stage) repo with `foo:bar:1.0.1:jar:bugfix-barf`.

* On branch `feature/poc` from `/develop`
    ```
    <version>1.1.0-SNAPSHOT</version>
    ```
    Deploys artifacts to the snapshots repo with 'foo:bar:1.1.0-SNAPSHOT:jar:feature/poc`.

This strategy would:
* address the use-case of other projects needing to add dependencies to long-running feature branches, originally called out in #83 by @robth. Those projects could depend upon the <classifier> versions.
* Protect the snapshots and test releases (stage) artifacts from being overwritten by branch builds. Enabling CD pipelines to spin-up environments deploying specific branch artifacts, while making sure those environments are not deploying yet-to-be-merged changes from the feature / bugfix branches.
* Allow us to remove the -SNAPSHOT check on feature / bugfix branches.
* Remove the complex checking and assertions around deploying snapshot artifacts with branch names in the version.

**Comment by bvarner** [comment]
From the looks of it, the only reason the snapshots check and FEATURE_OR_BUGFIX_BRANCH enum value was added, was to support enforcement of -SNAPSHOT on those branches, and to allow for those branches to be uploaded to (and fetched from) the snapshots repository.

I can understand the use case for wanting artifacts to go to the remote repo for specific long-running POC branches in projects that are managed separately but interdependent. I've been there before.
In previous versions of the plugin, these branches were **never** uploaded to the repos. Only dev, master, support, hotfix, and release branches were deployed to the repositories.

I can also see from some CI / CD perspectives why some may want to retrieve artifacts from the repositories from these branches.

But I don't think that should be the default behavior.

I think the default should be to opt-in for publishing non-artifact deploying branches to deploy to repositories. I also believe there should be some heuristics around how things get named / versioned in that case, and that these things should be able to be overridden.

**Comment by bvarner** [comment]
So what we settled on in 2.0.0 was addressed in #91.

To summarize the outcome of that PR here:

```
Introduces otherDeploymentBranchPattern, a regex that can match against any non-versioned (other, bugfix, feature, etc.) branch name pattern to force the following:

Automatic injection of the branch name (normalized to conform with semVer) into the resultant artifacts version string.
Forced addition of -SNAPSHOT to the artifacts version strings.
Deployment of these snapshot artifacts to the Snapshots repository.
This should fulfill the spirit of the changes that were added in 1.8.0 as a consequence of merging #83 and #84, but in a slightly more configurable manner, and without the negative impact on the feature and bugfix branch version enforcement called out in #87
```
